// node-cron/dist/tasks/inline-scheduled-task.js
import { EventEmitter } from 'events';
import { Runner } from '../scheduler/runner.js';
import { TimeMatcher } from '../time/time-matcher.js';
import { createID } from '../create-id.js';
import { StateMachine } from './state-machine.js';
import logger from '../logger.js';
import { LocalizedTime } from '../time/localized-time.js';
import { resolveRunCoordinator } from '../coordinator/run-coordinator.js';

class TaskEmitter extends EventEmitter {
}
class InlineScheduledTask {
    emitter;
    cronExpression;
    timeMatcher;
    runner;
    id;
    name;
    stateMachine;
    timezone;
    logger;
    suppressMissedWarning;
    constructor(cronExpression, taskFn, options) {
        this.emitter = new TaskEmitter();
        this.cronExpression = cronExpression;
        this.id = createID('task', 12);
        this.name = options?.name || this.id;
        this.timezone = options?.timezone;
        this.logger = options?.logger || logger;
        this.suppressMissedWarning = options?.suppressMissedWarning || false;
        this.timeMatcher = new TimeMatcher(cronExpression, options?.timezone);
        this.stateMachine = new StateMachine();
        const runnerOptions = {
            timezone: options?.timezone,
            noOverlap: options?.noOverlap,
            maxExecutions: options?.maxExecutions,
            maxRandomDelay: options?.maxRandomDelay,
            missedExecutionTolerance: options?.missedExecutionTolerance,
            logger: this.logger,
            beforeRun: (date, execution) => {
                if (execution.reason === 'scheduled') {
                    this.changeState('running');
                }
                this.emitter.emit('execution:started', this.createContext(date, execution));
                return true;
            },
            onFinished: (date, execution) => {
                if (execution.reason === 'scheduled') {
                    this.changeState('idle');
                }
                this.emitter.emit('execution:finished', this.createContext(date, execution));
                return true;
            },
            onError: (date, error, execution) => {
                this.logger.error(error);
                this.emitter.emit('execution:failed', this.createContext(date, execution));
                this.changeState('idle');
            },
            onOverlap: (date) => {
                this.emitter.emit('execution:overlap', this.createContext(date));
            },
            onMissedExecution: (date) => {
                const handled = this.emitter.listenerCount('execution:missed') > 0;
                if (!this.suppressMissedWarning && !handled) {
                    this.logger.warn(`missed execution at ${date}! Possible blocking IO or high CPU user at the same process used by node-cron.`);
                }
                this.emitter.emit('execution:missed', this.createContext(date));
            },
            onMaxExecutions: (date) => {
                this.emitter.emit('execution:maxReached', this.createContext(date));
                this.destroy();
            },
            runCoordinator: options?.distributed ? resolveRunCoordinator(options?.runCoordinator) : undefined,
            coordinatorKeyPrefix: this.name,
            coordinatorTtl: options?.distributedLease,
            onSkipped: (date, reason) => {
                this.emitter.emit('execution:skipped', this.createContext(date, undefined, reason));
            }
        };
        this.runner = new Runner(this.timeMatcher, (date, execution) => {
            return taskFn(this.createContext(date, execution));
        }, runnerOptions);
    }
    getNextRun() {
        if (this.stateMachine.state !== 'stopped') {
            return this.runner.nextRun();
        }
        return null;
    }
    getNextRuns(count) {
        const runs = [];
        let from = new Date();
        for (let i = 0; i < count; i++) {
            from = this.timeMatcher.getNextMatch(from);
            runs.push(from);
        }
        return runs;
    }
    match(date) {
        return this.timeMatcher.match(date);
    }
    msToNext() {
        const next = this.getNextRun();
        return next ? next.getTime() - Date.now() : null;
    }
    isBusy() {
        return this.getStatus() === 'running';
    }
    runsLeft() {
        if (this.runner.maxExecutions == null)
            return undefined;
        return Math.max(0, this.runner.maxExecutions - this.runner.runCount);
    }
    getPattern() {
        return this.cronExpression;
    }
    changeState(state) {
        if (this.runner.isStarted()) {
            this.stateMachine.changeState(state);
        }
    }
    start() {
        if (this.runner.isStopped()) {
            this.runner.start();
            this.stateMachine.changeState('idle');
            this.emitter.emit('task:started', this.createContext(new Date()));
        }
    }
    stop() {
        if (this.runner.isStarted()) {
            this.runner.stop();
            this.stateMachine.changeState('stopped');
            this.emitter.emit('task:stopped', this.createContext(new Date()));
        }
    }
    getStatus() {
        return this.stateMachine.state;
    }
    destroy() {
        if (this.stateMachine.state === 'destroyed')
            return;
        this.stop();
        this.stateMachine.changeState('destroyed');
        this.emitter.emit('task:destroyed', this.createContext(new Date()));
    }
    execute() {
        return new Promise((resolve, reject) => {
            const onFail = (context) => {
                this.off('execution:finished', onFinished);
                reject(context.execution?.error);
            };
            const onFinished = (context) => {
                this.off('execution:failed', onFail);
                resolve(context.execution?.result);
            };
            this.once('execution:finished', onFinished);
            this.once('execution:failed', onFail);
            this.runner.execute();
        });
    }
    on(event, fun) {
        this.emitter.on(event, fun);
    }
    off(event, fun) {
        this.emitter.off(event, fun);
    }
    once(event, fun) {
        this.emitter.once(event, fun);
    }
    createContext(executionDate, execution, reason) {
        const localTime = new LocalizedTime(executionDate, this.timezone);
        const ctx = {
            date: localTime.toDate(),
            dateLocalIso: localTime.toISO(),
            triggeredAt: new Date(),
            task: this,
            execution: execution
        };
        if (reason)
            ctx.reason = reason;
        return ctx;
    }
}

export { InlineScheduledTask };
