// node-cron/dist/pattern/validation/pattern-validation.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var index = require('../conversion/index.cjs');

const validationRegex = /^(?:\d+|\*|\*\/\d+)$/;
function isValidExpression(expression, min, max) {
    const options = expression;
    for (const option of options) {
        const optionAsInt = parseInt(option, 10);
        if ((!Number.isNaN(optionAsInt) &&
            (optionAsInt < min || optionAsInt > max)) ||
            !validationRegex.test(option))
            return false;
    }
    return true;
}
function isInvalidSecond(expression) {
    return !isValidExpression(expression, 0, 59);
}
function isInvalidMinute(expression) {
    return !isValidExpression(expression, 0, 59);
}
function isInvalidHour(expression) {
    return !isValidExpression(expression, 0, 23);
}
function isInvalidDayOfMonth(expression) {
    const days = expression.filter((value) => value !== 'L');
    return !isValidExpression(days, 1, 31);
}
function isInvalidMonth(expression) {
    return !isValidExpression(expression, 1, 12);
}
function isInvalidWeekDay(expression) {
    return !isValidExpression(expression, 0, 7);
}
function validateFields(patterns, executablePatterns) {
    if (isInvalidSecond(executablePatterns[0]))
        throw new Error(`${patterns[0]} is a invalid expression for second`);
    if (isInvalidMinute(executablePatterns[1]))
        throw new Error(`${patterns[1]} is a invalid expression for minute`);
    if (isInvalidHour(executablePatterns[2]))
        throw new Error(`${patterns[2]} is a invalid expression for hour`);
    if (isInvalidDayOfMonth(executablePatterns[3]))
        throw new Error(`${patterns[3]} is a invalid expression for day of month`);
    if (isInvalidMonth(executablePatterns[4]))
        throw new Error(`${patterns[4]} is a invalid expression for month`);
    if (isInvalidWeekDay(executablePatterns[5]))
        throw new Error(`${patterns[5]} is a invalid expression for week day`);
}
const FIELDS = [
    { key: 'second', label: 'second', invalid: isInvalidSecond },
    { key: 'minute', label: 'minute', invalid: isInvalidMinute },
    { key: 'hour', label: 'hour', invalid: isInvalidHour },
    { key: 'dayOfMonth', label: 'day of month', invalid: isInvalidDayOfMonth },
    { key: 'month', label: 'month', invalid: isInvalidMonth },
    { key: 'dayOfWeek', label: 'week day', invalid: isInvalidWeekDay },
];
function validateDetailed(pattern) {
    if (typeof pattern !== 'string')
        return { valid: false, errors: [{ field: 'expression', message: 'pattern must be a string' }] };
    if (!/^[a-zA-Z0-9-*/, ]+$/.test(pattern))
        return { valid: false, errors: [{ field: 'expression', value: pattern, message: 'pattern includes illegal characters' }] };
    const raw = pattern.replace(/\s{2,}/g, ' ').trim().split(' ');
    if (raw.length !== 5 && raw.length !== 6)
        return { valid: false, errors: [{ field: 'expression', value: pattern, message: `expected 5 or 6 fields but got ${raw.length}` }] };
    const patterns = raw.length === 5 ? ['0', ...raw] : raw;
    const executable = index.default(pattern);
    const errors = [];
    FIELDS.forEach((f, i) => {
        if (f.invalid(executable[i]))
            errors.push({ field: f.key, value: patterns[i], message: `${patterns[i]} is a invalid expression for ${f.label}` });
    });
    if (errors.length)
        return { valid: false, errors };
    return {
        valid: true,
        errors: [],
        fields: {
            second: executable[0],
            minute: executable[1],
            hour: executable[2],
            dayOfMonth: executable[3],
            month: executable[4],
            dayOfWeek: executable[5],
        },
    };
}
function parse(pattern) {
    const result = validateDetailed(pattern);
    if (!result.valid)
        throw new Error(result.errors[0].message);
    return result.fields;
}
function validate(pattern) {
    if (typeof pattern !== 'string')
        throw new TypeError('pattern must be a string!');
    const charRegex = new RegExp('^[a-zA-Z0-9-*/, ]+$');
    if (!charRegex.test(pattern))
        throw new TypeError('pattern includes illegal characters!');
    const patterns = pattern.split(' ');
    const executablePatterns = index.default(pattern);
    if (patterns.length === 5)
        patterns.unshift('0');
    validateFields(patterns, executablePatterns);
}

exports.default = validate;
exports.parse = parse;
exports.validateDetailed = validateDetailed;
