// node-cron/dist/time/localized-time.cjs
'use strict';

class LocalizedTime {
    timestamp;
    parts;
    timezone;
    constructor(date, timezone) {
        this.timestamp = date.getTime();
        this.timezone = timezone;
        this.parts = buildDateParts(date, timezone);
    }
    toDate() {
        return new Date(this.timestamp);
    }
    toISO() {
        const gmt = this.parts.gmt.replace(/^GMT/, '');
        const offset = gmt ? gmt : 'Z';
        const pad = (n) => String(n).padStart(2, '0');
        return `${this.parts.year}-${pad(this.parts.month)}-${pad(this.parts.day)}`
            + `T${pad(this.parts.hour)}:${pad(this.parts.minute)}:${pad(this.parts.second)}`
            + `.${String(this.parts.millisecond).padStart(3, '0')}`
            + offset;
    }
    getParts() {
        return this.parts;
    }
}
function getOffsetMinutes(date, timezone) {
    const offset = parseOffsetMinutes(getTimezoneGMT(date, timezone).replace(/^GMT/, '') || 'Z');
    return offset ?? 0;
}
function readsBackTo(timestamp, parts, timezone) {
    const p = buildDateParts(new Date(timestamp), timezone);
    return p.year === parts.year && p.month === parts.month && p.day === parts.day
        && p.hour === parts.hour && p.minute === parts.minute && p.second === parts.second;
}
function localTimeToTimestamp(parts, timezone) {
    const guess = Date.UTC(parts.year, parts.month - 1, parts.day, parts.hour, parts.minute, parts.second, parts.millisecond);
    const firstOffset = getOffsetMinutes(new Date(guess), timezone);
    const candidate1 = guess - firstOffset * 60000;
    const secondOffset = getOffsetMinutes(new Date(candidate1), timezone);
    if (secondOffset === firstOffset) {
        return candidate1;
    }
    const candidate2 = guess - secondOffset * 60000;
    if (readsBackTo(candidate1, parts, timezone))
        return candidate1;
    if (readsBackTo(candidate2, parts, timezone))
        return candidate2;
    return Math.max(candidate1, candidate2);
}
function buildDateParts(date, timezone) {
    const dftOptions = {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
        hour: '2-digit',
        minute: '2-digit',
        second: '2-digit',
        weekday: 'short',
        hour12: false
    };
    if (timezone) {
        dftOptions.timeZone = timezone;
    }
    const dateFormat = new Intl.DateTimeFormat('en-US', dftOptions);
    const parts = dateFormat.formatToParts(date).filter(part => {
        return part.type !== 'literal';
    }).reduce((acc, part) => {
        acc[part.type] = part.value;
        return acc;
    }, {});
    return {
        day: parseInt(parts.day),
        month: parseInt(parts.month),
        year: parseInt(parts.year),
        hour: parts.hour === '24' ? 0 : parseInt(parts.hour),
        minute: parseInt(parts.minute),
        second: parseInt(parts.second),
        millisecond: date.getMilliseconds(),
        weekday: parts.weekday,
        gmt: getTimezoneGMT(date, timezone)
    };
}
function parseOffsetMinutes(isoString) {
    if (isoString.endsWith('Z'))
        return 0;
    const match = isoString.match(/([+-])(\d{2}):(\d{2})$/);
    if (!match)
        return null;
    const sign = match[1] === '+' ? 1 : -1;
    return sign * (parseInt(match[2]) * 60 + parseInt(match[3]));
}
function getTimezoneGMT(date, timezone) {
    const fmt = new Intl.DateTimeFormat('en-US', {
        timeZone: timezone,
        timeZoneName: 'shortOffset'
    });
    const parts = fmt.formatToParts(date);
    const tzPart = parts.find(p => p.type === 'timeZoneName');
    if (!tzPart)
        return 'Z';
    const tzValue = tzPart.value;
    if (tzValue === 'GMT')
        return 'Z';
    const match = tzValue.match(/^GMT([+-])(\d{1,2})(?::(\d{2}))?$/);
    if (!match)
        return 'Z';
    const sign = match[1];
    const hoursNum = parseInt(match[2]);
    const minutesNum = parseInt(match[3] || '0');
    if (hoursNum === 0 && minutesNum === 0)
        return 'Z';
    const hours = match[2].padStart(2, '0');
    const minutes = (match[3] || '00').padStart(2, '0');
    return `GMT${sign}${hours}:${minutes}`;
}

exports.LocalizedTime = LocalizedTime;
exports.localTimeToTimestamp = localTimeToTimestamp;
