// node-cron/dist/time/matcher-walker.cjs
'use strict';

var index = require('../pattern/conversion/index.cjs');
var localizedTime = require('./localized-time.cjs');
var timeMatcher = require('./time-matcher.cjs');
var dayOfMonth = require('./day-of-month.cjs');

const MAX_DAYS = 366 * 100;
class MatcherWalker {
    cronExpression;
    baseDate;
    timeMatcher;
    timezone;
    seconds;
    minutes;
    hours;
    days;
    months;
    weekdays;
    constructor(cronExpression, baseDate, timezone) {
        this.cronExpression = cronExpression;
        this.baseDate = baseDate;
        this.timeMatcher = new timeMatcher.TimeMatcher(cronExpression, timezone);
        this.timezone = timezone;
        const expressions = index.default(cronExpression);
        this.seconds = sortedAsc(expressions[0]);
        this.minutes = sortedAsc(expressions[1]);
        this.hours = sortedAsc(expressions[2]);
        this.days = expressions[3];
        this.months = expressions[4];
        this.weekdays = expressions[5];
    }
    isMatching() {
        return this.timeMatcher.match(this.baseDate);
    }
    matchNext() {
        const months = this.months;
        const days = this.days;
        const baseMs = Math.floor(this.baseDate.getTime() / 1000) * 1000;
        const baseParts = new localizedTime.LocalizedTime(new Date(baseMs), this.timezone).getParts();
        let { year, month, day } = baseParts;
        for (let i = 0; i < MAX_DAYS; i++) {
            if (months.includes(month) && dayOfMonth.matchesDayOfMonth(days, year, month, day) && this.matchesWeekday(year, month, day)) {
                const lowerBound = i === 0 ? baseParts : null;
                const found = this.firstTimeOnDay(year, month, day, lowerBound, baseMs);
                if (found !== null) {
                    return new localizedTime.LocalizedTime(new Date(found), this.timezone);
                }
            }
            ({ year, month, day } = nextDay(year, month, day));
        }
        throw new Error('Could not find next matching date within reasonable time range');
    }
    firstTimeOnDay(year, month, day, lowerBound, baseMs) {
        const { seconds, minutes, hours } = this;
        for (const hour of hours) {
            if (lowerBound && hour < lowerBound.hour)
                continue;
            for (const minute of minutes) {
                for (const second of seconds) {
                    if (lowerBound && !isLaterInDay(hour, minute, second, lowerBound))
                        continue;
                    const ts = localizedTime.localTimeToTimestamp({ year, month, day, hour, minute, second, millisecond: 0 }, this.timezone);
                    if (ts > baseMs && this.timeMatcher.match(new Date(ts))) {
                        return ts;
                    }
                }
            }
        }
        return null;
    }
    matchesWeekday(year, month, day) {
        const weekday = new Date(Date.UTC(year, month - 1, day)).getUTCDay();
        return this.weekdays.indexOf(weekday) !== -1;
    }
}
function nextDay(year, month, day) {
    const d = new Date(Date.UTC(year, month - 1, day + 1));
    return { year: d.getUTCFullYear(), month: d.getUTCMonth() + 1, day: d.getUTCDate() };
}
function sortedAsc(values) {
    return [...values].sort((a, b) => a - b);
}
function isLaterInDay(hour, minute, second, bound) {
    return hour * 3600 + minute * 60 + second > bound.hour * 3600 + bound.minute * 60 + bound.second;
}

exports.MatcherWalker = MatcherWalker;
