// node-cron/dist/time/time-matcher.cjs
'use strict';

var index = require('../pattern/conversion/index.cjs');
var weekDayNamesConversion = require('../pattern/conversion/week-day-names-conversion.cjs');
var localizedTime = require('./localized-time.cjs');
var matcherWalker = require('./matcher-walker.cjs');
var dayOfMonth = require('./day-of-month.cjs');

function matchValue(allowedValues, value) {
    return allowedValues.indexOf(value) !== -1;
}
class TimeMatcher {
    timezone;
    pattern;
    expressions;
    constructor(pattern, timezone) {
        this.timezone = timezone;
        this.pattern = pattern;
        this.expressions = index.default(pattern);
    }
    match(date) {
        const localizedTime$1 = new localizedTime.LocalizedTime(date, this.timezone);
        const parts = localizedTime$1.getParts();
        const runOnSecond = matchValue(this.expressions[0], parts.second);
        const runOnMinute = matchValue(this.expressions[1], parts.minute);
        const runOnHour = matchValue(this.expressions[2], parts.hour);
        const runOnDay = dayOfMonth.matchesDayOfMonth(this.expressions[3], parts.year, parts.month, parts.day);
        const runOnMonth = matchValue(this.expressions[4], parts.month);
        const runOnWeekDay = matchValue(this.expressions[5], parseInt(weekDayNamesConversion.default(parts.weekday)));
        return runOnSecond && runOnMinute && runOnHour && runOnDay && runOnMonth && runOnWeekDay;
    }
    getNextMatch(date) {
        const walker = new matcherWalker.MatcherWalker(this.pattern, date, this.timezone);
        const next = walker.matchNext();
        return next.toDate();
    }
}

exports.TimeMatcher = TimeMatcher;
