// node-cron/dist/tasks/background-scheduled-task/background-scheduled-task.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');
require('url');
var child_process = require('child_process');
var createId = require('../../create-id.cjs');
var events = require('events');
var stateMachine = require('../state-machine.cjs');
var localizedTime = require('../../time/localized-time.cjs');
var logger = require('../../logger.cjs');
var timeMatcher = require('../../time/time-matcher.cjs');
var runCoordinator = require('../../coordinator/run-coordinator.cjs');

const daemonPath = path.resolve(path.dirname(__filename), 'daemon.cjs');
class TaskEmitter extends events.EventEmitter {
}
class BackgroundScheduledTask {
    emitter;
    id;
    name;
    cronExpression;
    taskPath;
    options;
    forkProcess;
    stateMachine;
    logger;
    suppressMissedWarning;
    timeMatcher;
    runCount;
    runCoordinator;
    constructor(cronExpression, taskPath, options) {
        this.cronExpression = cronExpression;
        this.taskPath = taskPath;
        this.options = options;
        this.id = createId.createID('task');
        this.name = options?.name || this.id;
        this.emitter = new TaskEmitter();
        this.stateMachine = new stateMachine.StateMachine('stopped');
        this.timeMatcher = new timeMatcher.TimeMatcher(cronExpression, options?.timezone);
        this.runCount = 0;
        this.on('execution:started', () => { this.runCount++; });
        this.logger = options?.logger || logger.default;
        this.suppressMissedWarning = options?.suppressMissedWarning || false;
        this.runCoordinator = options?.distributed ? runCoordinator.resolveRunCoordinator(options?.runCoordinator) : undefined;
        this.on('task:stopped', () => {
            this.forkProcess?.kill();
            this.forkProcess = undefined;
            this.stateMachine.changeState('stopped');
        });
        this.on('task:destroyed', () => {
            this.forkProcess?.kill();
            this.forkProcess = undefined;
            this.stateMachine.changeState('destroyed');
        });
    }
    getNextRun() {
        if (this.stateMachine.state !== 'stopped') {
            return this.timeMatcher.getNextMatch(new Date());
        }
        return null;
    }
    getNextRuns(count) {
        const runs = [];
        let from = new Date();
        for (let i = 0; i < count; i++) {
            from = this.timeMatcher.getNextMatch(from);
            runs.push(from);
        }
        return runs;
    }
    match(date) {
        return this.timeMatcher.match(date);
    }
    msToNext() {
        const next = this.getNextRun();
        return next ? next.getTime() - Date.now() : null;
    }
    isBusy() {
        return this.getStatus() === 'running';
    }
    runsLeft() {
        if (this.options?.maxExecutions == null)
            return undefined;
        return Math.max(0, this.options.maxExecutions - this.runCount);
    }
    getPattern() {
        return this.cronExpression;
    }
    start() {
        return new Promise((resolve, reject) => {
            if (this.forkProcess) {
                return resolve(undefined);
            }
            const startTimeout = this.options?.startTimeout ?? 5000;
            const failStart = (error) => {
                clearTimeout(timeout);
                this.forkProcess?.kill();
                this.forkProcess = undefined;
                reject(error);
            };
            const timeout = setTimeout(() => {
                failStart(new Error(`Start operation timed out after ${startTimeout}ms. The background task file may have failed to load or taken too long to import; ` +
                    `verify it runs on its own and consider increasing the \`startTimeout\` option.`));
            }, startTimeout);
            try {
                this.forkProcess = child_process.fork(daemonPath);
                this.forkProcess.on('error', (err) => {
                    failStart(new Error(`Error on daemon: ${err.message}`));
                });
                this.forkProcess.on('exit', (code, signal) => {
                    if (code !== 0 && signal !== 'SIGTERM') {
                        const erro = new Error(`node-cron daemon exited with code ${code || signal}`);
                        this.logger.error(erro);
                        failStart(erro);
                    }
                });
                this.forkProcess.on('message', (message) => {
                    if (message.type === 'coordinator:shouldRun') {
                        void this.handleShouldRun(message);
                        return;
                    }
                    if (message.type === 'coordinator:complete') {
                        this.runCoordinator?.onComplete?.(message.key)?.catch?.((err) => this.logger.error('Run coordinator onComplete failed', err));
                        return;
                    }
                    if (message.event === 'daemon:error') {
                        failStart(message.jsonError ? deserializeError(message.jsonError) : new Error('Background task failed to start'));
                        return;
                    }
                    if (message.jsonError) {
                        if (message.context?.execution) {
                            message.context.execution.error = deserializeError(message.jsonError);
                            delete message.jsonError;
                        }
                    }
                    if (message.context?.task?.state) {
                        this.stateMachine.changeState(message.context?.task?.state);
                    }
                    if (message.context) {
                        const execution = message.context?.execution;
                        delete execution?.hasError;
                        const context = this.createContext(new Date(message.context.date), execution, message.context.reason);
                        this.logEvent(message.event, context);
                        this.emitter.emit(message.event, context);
                    }
                });
                this.once('task:started', () => {
                    this.stateMachine.changeState('idle');
                    clearTimeout(timeout);
                    resolve(undefined);
                });
                this.forkProcess.send({
                    command: 'task:start',
                    path: this.taskPath,
                    cron: this.cronExpression,
                    options: serializableOptions(this.options)
                });
            }
            catch (error) {
                failStart(error);
            }
        });
    }
    stop() {
        return new Promise((resolve, reject) => {
            if (!this.forkProcess) {
                return resolve(undefined);
            }
            const timeoutId = setTimeout(() => {
                clearTimeout(timeoutId);
                reject(new Error('Stop operation timed out'));
            }, 5000);
            const cleanupAndResolve = () => {
                clearTimeout(timeoutId);
                this.off('task:stopped', onStopped);
                this.forkProcess = undefined;
                resolve(undefined);
            };
            const onStopped = () => {
                cleanupAndResolve();
            };
            this.once('task:stopped', onStopped);
            this.forkProcess.send({
                command: 'task:stop'
            });
        });
    }
    getStatus() {
        return this.stateMachine.state;
    }
    destroy() {
        return new Promise((resolve, reject) => {
            if (!this.forkProcess) {
                return resolve(undefined);
            }
            const timeoutId = setTimeout(() => {
                clearTimeout(timeoutId);
                reject(new Error('Destroy operation timed out'));
            }, 5000);
            const onDestroy = () => {
                clearTimeout(timeoutId);
                this.off('task:destroyed', onDestroy);
                resolve(undefined);
            };
            this.once('task:destroyed', onDestroy);
            this.forkProcess.send({
                command: 'task:destroy'
            });
        });
    }
    execute() {
        return new Promise((resolve, reject) => {
            if (!this.forkProcess) {
                return reject(new Error('Cannot execute background task because it hasn\'t been started yet. Please initialize the task using the start() method before attempting to execute it.'));
            }
            let timeoutId;
            if (typeof this.options?.executeTimeout === 'number') {
                timeoutId = setTimeout(() => {
                    cleanupListeners();
                    reject(new Error('Execution timeout exceeded'));
                }, this.options.executeTimeout);
            }
            const cleanupListeners = () => {
                if (timeoutId)
                    clearTimeout(timeoutId);
                this.off('execution:finished', onFinished);
                this.off('execution:failed', onFail);
            };
            const onFinished = (context) => {
                cleanupListeners();
                resolve(context.execution?.result);
            };
            const onFail = (context) => {
                cleanupListeners();
                reject(context.execution?.error || new Error('Execution failed without specific error'));
            };
            this.once('execution:finished', onFinished);
            this.once('execution:failed', onFail);
            this.forkProcess.send({
                command: 'task:execute'
            });
        });
    }
    async handleShouldRun(message) {
        let allowed = false;
        let error;
        try {
            allowed = this.runCoordinator ? await this.runCoordinator.shouldRun(message.key, message.ttlMs) : false;
        }
        catch (err) {
            error = err?.message ?? String(err);
        }
        this.forkProcess?.send({ type: 'coordinator:result', reqId: message.reqId, allowed, error });
    }
    on(event, fun) {
        this.emitter.on(event, fun);
    }
    off(event, fun) {
        this.emitter.off(event, fun);
    }
    once(event, fun) {
        this.emitter.once(event, fun);
    }
    logEvent(event, context) {
        switch (event) {
            case 'execution:missed': {
                const handled = this.emitter.listenerCount('execution:missed') > 0;
                if (!this.suppressMissedWarning && !handled) {
                    this.logger.warn(`missed execution at ${context.date}! Possible blocking IO or high CPU user at the same process used by node-cron.`);
                }
                break;
            }
            case 'execution:overlap':
                if (this.options?.noOverlap) {
                    this.logger.warn('task still running, new execution blocked by overlap prevention!');
                }
                break;
            case 'execution:failed':
                if (context.execution?.error) {
                    this.logger.error(context.execution.error);
                }
                break;
        }
    }
    createContext(executionDate, execution, reason) {
        const localTime = new localizedTime.LocalizedTime(executionDate, this.options?.timezone);
        const ctx = {
            date: localTime.toDate(),
            dateLocalIso: localTime.toISO(),
            triggeredAt: new Date(),
            task: this,
            execution: execution
        };
        if (reason)
            ctx.reason = reason;
        return ctx;
    }
}
function serializableOptions(options) {
    if (!options)
        return options;
    const { logger: _logger, runCoordinator: _runCoordinator, ...rest } = options;
    return rest;
}
function deserializeError(str) {
    const data = JSON.parse(str);
    const Err = globalThis[data.name] || Error;
    const err = new Err(data.message);
    if (data.stack) {
        err.stack = data.stack;
    }
    Object.keys(data).forEach(key => {
        if (!['name', 'message', 'stack'].includes(key)) {
            err[key] = data[key];
        }
    });
    return err;
}

exports.default = BackgroundScheduledTask;
