// node-cron/dist/scheduler/runner.js
import { createID } from '../create-id.js';
import logger from '../logger.js';
import { TrackedPromise } from '../promise/tracked-promise.js';
import { planBeat } from './plan-beat.js';

const DEFAULT_MISSED_EXECUTION_TOLERANCE = 1000;
function emptyOnFn() { }
function emptySkipFn() { }
function emptyHookFn() { return true; }
const DEFAULT_COORDINATOR_TTL = 30000;
class Runner {
    timeMatcher;
    onMatch;
    noOverlap;
    maxExecutions;
    maxRandomDelay;
    missedExecutionTolerance;
    runCount;
    running;
    heartBeatTimeout;
    logger;
    onMissedExecution;
    onOverlap;
    onError;
    beforeRun;
    onFinished;
    onMaxExecutions;
    runCoordinator;
    coordinatorKeyPrefix;
    coordinatorTtl;
    onSkipped;
    constructor(timeMatcher, onMatch, options) {
        this.timeMatcher = timeMatcher;
        this.onMatch = onMatch;
        this.noOverlap = options == undefined || options.noOverlap === undefined ? false : options.noOverlap;
        this.maxExecutions = options?.maxExecutions;
        this.maxRandomDelay = options?.maxRandomDelay || 0;
        this.missedExecutionTolerance = options?.missedExecutionTolerance ?? DEFAULT_MISSED_EXECUTION_TOLERANCE;
        this.logger = options?.logger || logger;
        this.onMissedExecution = options?.onMissedExecution || emptyOnFn;
        this.onOverlap = options?.onOverlap || emptyOnFn;
        this.onError = options?.onError || ((date, error) => this.logger.error('Task failed with error!', error));
        this.onFinished = options?.onFinished || emptyHookFn;
        this.beforeRun = options?.beforeRun || emptyHookFn;
        this.onMaxExecutions = options?.onMaxExecutions || emptyOnFn;
        this.runCoordinator = options?.runCoordinator;
        this.coordinatorKeyPrefix = options?.coordinatorKeyPrefix || '';
        this.coordinatorTtl = options?.coordinatorTtl ?? DEFAULT_COORDINATOR_TTL;
        this.onSkipped = options?.onSkipped || emptySkipFn;
        this.runCount = 0;
        this.running = false;
    }
    onErrorFallback = (date, error) => {
        this.logger.error('Task failed with error!', error);
    };
    async runCoordinated(slot, run) {
        if (!this.runCoordinator) {
            await run();
            return;
        }
        const key = `${this.coordinatorKeyPrefix}:${slot.toISOString()}`;
        let allowed;
        try {
            allowed = await this.runCoordinator.shouldRun(key, this.coordinatorTtl);
        }
        catch (err) {
            this.logger.error('Run coordinator failed; skipping execution (fail-closed)', err);
            this.emitSkipped(slot, 'coordinator-error');
            return;
        }
        if (!allowed) {
            this.emitSkipped(slot, 'not-elected');
            return;
        }
        try {
            await run();
        }
        finally {
            try {
                await this.runCoordinator.onComplete?.(key);
            }
            catch (err) {
                this.logger.error('Run coordinator onComplete failed', err);
            }
        }
    }
    emitSkipped(slot, reason) {
        Promise.resolve(this.onSkipped(slot, reason)).catch((err) => this.onErrorFallback(slot, err));
    }
    start() {
        this.running = true;
        let lastExecution;
        let expectedNextExecution = this.timeMatcher.getNextMatch(nowWithoutMs());
        const armHeartBeat = () => {
            if (this.running) {
                clearTimeout(this.heartBeatTimeout);
                this.heartBeatTimeout = setTimeout(heartBeat, getDelay(expectedNextExecution));
            }
        };
        const runTask = (date) => {
            return new Promise(async (resolve) => {
                const execution = {
                    id: createID('exec'),
                    reason: 'scheduled'
                };
                const shouldExecute = await this.beforeRun(date, execution);
                const randomDelay = Math.floor(Math.random() * this.maxRandomDelay);
                if (shouldExecute) {
                    setTimeout(async () => {
                        try {
                            this.runCount++;
                            execution.startedAt = new Date();
                            const result = await this.onMatch(date, execution);
                            execution.finishedAt = new Date();
                            execution.result = result;
                            this.onFinished(date, execution);
                            if (this.maxExecutions && this.runCount >= this.maxExecutions) {
                                this.onMaxExecutions(date);
                                this.stop();
                            }
                        }
                        catch (error) {
                            execution.finishedAt = new Date();
                            execution.error = error;
                            this.onError(date, error, execution);
                        }
                        resolve(true);
                    }, randomDelay);
                }
                else {
                    resolve(true);
                }
            });
        };
        const heartBeat = async () => {
            const currentDate = nowWithoutMs();
            const plan = planBeat(expectedNextExecution, currentDate, this.missedExecutionTolerance, (date) => this.timeMatcher.getNextMatch(date));
            expectedNextExecution = plan.next;
            for (const missedSlot of plan.missed) {
                runAsync(this.onMissedExecution, missedSlot, this.onErrorFallback);
            }
            if (plan.run) {
                if (lastExecution && lastExecution.getState() === 'pending') {
                    runAsync(this.onOverlap, plan.run, this.onErrorFallback);
                    if (this.noOverlap) {
                        this.logger.warn('task still running, new execution blocked by overlap prevention!');
                        armHeartBeat();
                        return;
                    }
                }
                const slot = plan.run;
                lastExecution = new TrackedPromise(async (resolve, reject) => {
                    try {
                        await this.runCoordinated(slot, () => runTask(slot));
                        resolve(true);
                    }
                    catch (err) {
                        reject(err);
                    }
                });
            }
            armHeartBeat();
        };
        armHeartBeat();
    }
    nextRun() {
        return this.timeMatcher.getNextMatch(new Date());
    }
    stop() {
        this.running = false;
        if (this.heartBeatTimeout) {
            clearTimeout(this.heartBeatTimeout);
            this.heartBeatTimeout = undefined;
        }
    }
    isStarted() {
        return !!this.heartBeatTimeout && this.running;
    }
    isStopped() {
        return !this.isStarted();
    }
    async execute() {
        const date = new Date();
        const execution = {
            id: createID('exec'),
            reason: 'invoked'
        };
        try {
            const shouldExecute = await this.beforeRun(date, execution);
            if (shouldExecute) {
                this.runCount++;
                execution.startedAt = new Date();
                const result = await this.onMatch(date, execution);
                execution.finishedAt = new Date();
                execution.result = result;
                this.onFinished(date, execution);
            }
        }
        catch (error) {
            execution.finishedAt = new Date();
            execution.error = error;
            this.onError(date, error, execution);
        }
    }
}
async function runAsync(fn, date, onError) {
    try {
        await fn(date);
    }
    catch (error) {
        onError(date, error);
    }
}
function getDelay(nextRun) {
    const maxDelay = 86400000;
    const now = new Date();
    const delay = nextRun.getTime() - now.getTime();
    if (delay > maxDelay) {
        return maxDelay;
    }
    return Math.max(0, delay);
}
function nowWithoutMs() {
    const date = new Date();
    date.setMilliseconds(0);
    return date;
}

export { Runner };
